File: 2/count-depth-increases-over-sliding-windows.rb
#!/usr/bin/env ruby

# See https://adventofcode.com/2021/day/2

# First calculate sums for slding windows.
input = File.readlines(ARGV[0]).map(&:to_i)

sliding_window_sums = input.each_cons(3).map(&:sum)

# Then calculate number of times depth increases
previous_sum = nil
number_of_times_depth_increases = 0

sliding_window_sums.each do |sum|
    number_of_times_depth_increases += 1 if previous_sum && sum > previous_sum

    previous_sum = sum
end

puts number_of_times_depth_increases

File: 2/count-depth-increases-over-sliding-windows.exs
# See https://adventofcode.com/2021/day/2
{:ok, depth_readings} = File.read("input.txt")
depth_readings = depth_readings |> String.split("\n", trim: true) |> Enum.map(&String.to_integer/1)

number_of_times_depth_increases = depth_readings
  |> Enum.chunk_every(3, 1, :discard)
  |> Enum.map(fn x -> Enum.sum(x) end)
  |> Enum.chunk_every(2, 1, :discard)
  |> Enum.map(fn x -> if List.last(x) > List.first(x), do: 1, else: 0 end)
  |> Enum.sum

IO.puts(number_of_times_depth_increases)
